### ISSUE TITLE
Slack backend doesn't process channel names from mobile properly

### ISSUE THREAD

**Issue by jasedit** [errbotio/errbot#816]
When processing statements in a plugin, it appears that the mobile client sends channel names slightly differently than the desktop client. From the desktop client, the logs show:
```
'text': '!say <#C0J9GN39C> Test',
```

while the mobile client shows:
```
'text': '!say <#C0J9GN39C|general> Test',
```

The first version parses to a channel reference, while the second parses to `general`, and thus is no longer a reference to a channel. I'm assuming the extra annotation is what is causing the problem, but I'm not able to spot where exactly in the code this is happening yet.

**Comment by nlochschmidt** [comment]
I have been looking into the Slack backend implementation to figure out whether I can quickly solve #812 and I am guessing your problem is the code that sanitizes urls:

https://github.com/errbotio/errbot/blob/53073eb721bc0e3a2eb0507a3bd80f8acd310d1a/errbot/backends/slack.py#L781-L794

So in your case the line `text = re.sub(r'<([^\|>]+)\|([^\|>]+)>', r'\2', text)` selects the group behind the `|` instead of the first group.

**Comment by jasedit** [comment]
That certainly looks like the place where the wrong group is selected, but from what I can see in the code, that function is used in many places to strip out URLs, and I'm loathe to change the line. I don't have a strong enough grasp of the overall system to know how mentions of rooms/users should be handled in the text processing to know if modifying `sanitize_urls` to retain them is correct, or if only specific functions should handle them.

**Comment by jasedit** [comment]
I've issued a PR that appears to solve this issue for me against the current master, and is due to how mentions are processed. The code did not correctly handle the particular form of channel identifiers sent by the mobile client, so instead of previous processing fixing the issue, it was passed to `sanitize_uri()`, which does not attempt to handle Slack specific references.

This bug still exists in the 4.2 branch, which has a fairly different block of code for reference parsing. I'm not sure if the bug should be addressed in the 4.2 branch, or if it's common practice to simply use the master branch/wait for another version release.

**Comment by zoni** [comment]
I believe this was fixed by #820, but please re-open if there are still any issues with this.

Also, for public record, master is our development branch so yes, PRs should generally go against that. We'll do the hard work for you and backport bugfixes to the branch for the current stable version if they're severe enough to warrant it. :smile:
